### PR TITLE
Allow customizing retry behavior for timeout failure

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
Allow customizing retry behavior for timeout errors
- NOTE: this changes the behavior for both activity and workflow retry policy

**Why?**
https://github.com/temporalio/temporal/issues/2496

**How did you test it?**
Added unit test

**Potential risks**
This change is technically breaking as the new format defined for timeout failure may matching user defined application error types. But the chance this will happen is very very low.

**Is hotfix candidate?**
no

### CODE REVIEW

File: service/history/workflow/retry.go [hunk 1]
```diff
@@ -131,16 +138,44 @@ func isRetryable(failure *failurepb.Failure, nonRetryableTypes []string) bool {
...
+}
+
+func matchApplicationNonRetryableTypes(
+	failureType string,
+	nonRetryableTypes []string,
+) bool {
+	for _, nrt := range nonRetryableTypes {
+		if strings.HasPrefix(nrt, common.TimeoutFailureTypePrefix) {
+			continue
+		}
```
Comment (yiminc) [inline service/history/workflow/retry.go:170]: no need for this check.

File: common/constants.go [hunk 2]
```diff
@@ -111,3 +111,11 @@ const (
 	// DefaultTransactionSizeLimit is the largest allowed transaction size to persistence
 	DefaultTransactionSizeLimit = 4 * 1024 * 1024
 )
+
+const (
+	// TimeoutFailureTypePrefix is the prefix for timeout failure types
+	// used in retry policy
+	// the actual failure type will be prefix + enums.TimeoutType.String()
+	// e.g. "Temporal Timeout: StartToClose" or "Temporal Timeout: Heartbeat"
+	TimeoutFailureTypePrefix = "Temporal Timeout: "
```
Comment (yiminc) [inline common/constants.go:120]: Maybe we should avoid space in the prefix?

**Review by yiminc — APPROVED** [review]

File: common/constants.go (hunk 2 above)
Comment (cretz) [reply common/constants.go:120]: :+1: Yup, this is a programmatic name not a sentence.

File: service/history/workflow/retry.go (hunk 1 above)
Comment (yycptt) [reply service/history/workflow/retry.go:170]: Removed

File: common/constants.go (hunk 2 above)
Comment (yycptt) [reply common/constants.go]: Sure. Removed both spaces, and now it becomes `TemporalTimeout:`

**Pull request merged**